File: examples/H20/robots/sdk_package/example_py/h10-w/status/get_controller_message.py
import humanoid_sdk_py
from humanoid_sdk_py import h10w


def parse_brake_state(brake_state):
    """解析抱闸状态"""
    print("抱闸状态 (32位):", bin(brake_state))
    print("抱闸状态详细解析:")

    # 关节抱闸状态解析
    print("  关节抱闸状态:")
    joint_names = [
        "L_JOINT_1",
        "L_JOINT_2",
        "L_JOINT_3",
        "L_JOINT_4",
        "L_JOINT_5",
        "L_JOINT_6",
        "L_JOINT_7",  # 0-6: 左臂关节
        "R_JOINT_1",
        "R_JOINT_2",
        "R_JOINT_3",
        "R_JOINT_4",
        "R_JOINT_5",
        "R_JOINT_6",
        "R_JOINT_7",  # 7-13: 右臂关节
        "HEAD_PITCH",
        "HEAD_YAW",  # 14-15: 头部关节
        "ELEVATOR_MOTOR",
        "MOBILE_BASE",  # 16-17: 升降电机和移动底盘
    ]

    for i in range(18):  # Bits 0-17 为有效位
        brake_engaged = (brake_state >> i) & 0x1
        joint_name = joint_names[i] if i < len(joint_names) else f"UNKNOWN_JOINT_{i}"
        status = "抱闸开启" if brake_engaged else "抱闸关闭"
        print(f"    {joint_name:<16}: {status}")

    # 预留位 (Bits 18-31)
    print("  预留位状态 (Bits 18-31):")
    for i in range(18, 32):
        bit_value = (brake_state >> i) & 0x1
        print(f"    Bit {i:2d}: {bit_value}")


def get_controller_status():
    """获取控制器状态信息"""
    print("=== 获取机器人当前控制器状态信息示例 ===")

    h10w_status = h10w.H10wStatus()
    ret, controller_msg = h10w_status.getControllerMessage()

    if ret != 0 or controller_msg is None:
        print(f"获取控制器信息失败，错误码: {ret}")
        return False

    print("\n=== 控制器状态信息 ===")

    # 控制器启用状态
    print("控制器启用状态: ", end="")
    if controller_msg.controller_enabled_state == 0:
        print("未启用")
    elif controller_msg.controller_enabled_state == 1:
        print("已启用")
    else:
        print(f"未知状态 ({controller_msg.controller_enabled_state})")

    # 电源状态
    print("机器人电源状态: ", end="")
    if controller_msg.power_state == 0:
        print("上电状态")
    elif controller_msg.power_state == 1:
        print("下电状态")
    else:
        print(f"未知状态 ({controller_msg.power_state})")

    # 抱闸状态
    parse_brake_state(controller_msg.brake_state)

    return True


get_controller_status()

File: examples/H20/robots/sdk_package/example/h10-w/status/get_controller_message.cpp
// get_controller_message.cpp

#include <iostream>
#include <iomanip>
#include <bitset>
#include "h10-w/h10w_status.h"
#include "common/common_def.h"

using namespace humanoid;
using namespace humanoid::h10w;

int main() {
  std::cout << "=== 获取机器人当前控制器状态信息示例 ===" << std::endl;

  H10wStatus status;
  ControllerMessageParams controllerMessage;

  int result = status.getControllerMessage(controllerMessage);
  if (result != 0) {
    std::cerr << "获取控制器信息失败，错误码: " << result << std::endl;
    return -1;
  }

  std::cout << "\n=== 控制器状态信息 ===" << std::endl;

  // 控制器启用状态
  std::cout << "控制器启用状态: ";
  if (controllerMessage.controller_enabled_state == 0) {
    std::cout << "未启用" << std::endl;
  } else if (controllerMessage.controller_enabled_state == 1) {
    std::cout << "已启用" << std::endl;
  } else {
    std::cout << "未知状态 ("
              << static_cast<int>(controllerMessage.controller_enabled_state)
              << ")" << std::endl;
  }

  // 电源状态
  std::cout << "机器人电源状态: ";
  if (controllerMessage.power_state == 0) {
    std::cout << "上电状态" << std::endl;
  } else if (controllerMessage.power_state == 1) {
    std::cout << "下电状态" << std::endl;
  } else {
    std::cout << "未知状态 (" << static_cast<int>(controllerMessage.power_state)
              << ")" << std::endl;
  }

  // 抱闸状态
  std::cout << "抱闸状态 (32位): "
            << std::bitset<32>(controllerMessage.brake_state) << std::endl;
  std::cout << "抱闸状态详细解析:" << std::endl;

  // 解析各关节抱闸状态
  std::cout << "  关节抱闸状态:" << std::endl;
  for (int i = 0; i < 18; i++) {  // Bits 0-17 为有效位
    bool brake_engaged = (controllerMessage.brake_state >> i) & 0x1;
    std::string joint_name;
    if (i < 7) {
      joint_name = "L_JOINT_" + std::to_string(i + 1);
    } else if (i < 14) {
      joint_name = "R_JOINT_" + std::to_string(i - 6);
    } else if (i == 14) {
      joint_name = "HEAD_PITCH";
    } else if (i == 15) {
      joint_name = "HEAD_YAW";
    } else if (i == 16) {
      joint_name = "ELEVATOR_MOTOR";
    } else if (i == 17) {
      joint_name = "MOBILE_BASE";
    } else {
      joint_name = "UNKNOWN_JOINT_" + std::to_string(i);
    }
    std::cout << "    " << std::left << std::setw(16) << joint_name << ": "
              << (brake_engaged ? "抱闸开启" : "抱闸关闭") << std::endl;
  }

  std::cout << "\n=== 状态信息获取完成 ===" << std::endl;
  return 0;
}